Added keyboard shortcut for main menu

dhruvipatel14 opened the pull request:
Added a new wagtail keyboard shortcut for Main menu with reference to this discussion: https://github.com/wagtail/wagtail/discussions/12050

Keyboard shortcut dialog after new change:

![Screenshot from 2024-12-12 16-57-03](https://github.com/user-attachments/assets/abd01877-d5f5-47a4-bcca-56ff5fde414f)

Unit test case status for javascript:

![Screenshot from 2024-12-12 17-31-22](https://github.com/user-attachments/assets/c3ad4be4-e409-4f76-a3b2-56df918950e2)

Let me know if I need to add a unit test for these changes, All existing test cases for keyboard shortcuts are passing.

lb- commented:
I think we should use just `[` (without the modifier), but I'll let others in the core team make a decision. 

@zerolab - in my absence, do you mind flagging this in the Core team chat? Would be good to get a direction for this PR. 

Thanks for raising @dhruvipatel14 - I'm away from my laptop for a month or so, I won't be able to fully review this. Thanks for raising though.

dhruvipatel14 commented:
Thank you for your feedback!
I’ve also opened another PR to document the shortcut for comments: https://github.com/wagtail/wagtail/pull/12679
I look forward to hearing back from the core team.

gasman commented:
Assigning to @thibaudcolas for accessibility considerations.

Scotchester commented:
> I think we should use just `[` (without the modifier), but I'll let others in the core team make a decision.

I would disagree with this, unless it is disabled when an input field is focused. If it worked even when an input is in focus, you'd be unable to intentionally type a `[` in your content.

lb- commented:
The code supports the ability to not trigger when on an input, if we do not set it to global I believe. 

So we can do the 'bare' shortcut if we want. It's what Trello and JIRA use (no need to use the modifier key), hence my suggestion.

Scotchester commented:
Gotcha, thanks for the clarification @lb-.

dhruvipatel14 commented:
Yes, that’s correct. If we remove the global scope, the shortcuts won’t trigger when an input field is focused.

dhruvipatel14 commented:
Hi @lb- , @thibaudcolas I wanted to follow up and ask if there’s any feedback on the keyboard shortcuts. Any updates or guidance would help me move forward with the changes.

lb- commented on client/src/components/Sidebar/Sidebar.tsx:
```diff
@@ -193,6 +193,9 @@ export const Sidebar: React.FunctionComponent<SidebarProps> = ({
         }
         type="button"
         ref={mobileNavToggleRef}
+        data-controller="w-kbd"
+        data-w-kbd-key-value="mod+["
+        data-w-kbd-scope-value="global"
```
Based on main comment, let's replace this with the 'target' attribute.

```suggestion
        data-w-kbd-target={isMobile ? 'element' : undefined}
```

Also, we should set the `aria-keyshortcuts` further up for completeness & accessibility.


```ts
  return (
    <>
      <button
        onClick={onClickOpenCloseToggle}
        aria-label={gettext('Toggle sidebar')}
        aria-expanded={visibleOnMobile ? 'true' : 'false'}
        aria-keyshortcuts="["
// ...
```

lb- commented on client/src/components/Sidebar/Sidebar.tsx:
```diff
@@ -233,6 +236,9 @@ export const Sidebar: React.FunctionComponent<SidebarProps> = ({
                 hover:opacity-100
                 more-contrast:w-border-border-interactive-more-contrast-dark-bg
                 hover:more-contrast:w-border-border-interactive-more-contrast-dark-bg-hover`}
+              data-controller="w-kbd"
+              data-w-kbd-key-value="mod+["
+              data-w-kbd-scope-value="global"
```
Same here, let's use the target approach, making it NOT the target when we are in mobile view.

```suggestion
              data-w-kbd-target={!isMobile ? 'element' : undefined}
```

And, let's set `aria-keyshortcuts` further up.

```ts
// ...
            <button
              onClick={onClickCollapseToggle}
              aria-label={gettext('Toggle sidebar')}
              aria-expanded={slim ? 'false' : 'true'}
              aria-keyshortcuts="["
              type="button"
// ...
```

lb- reviewed (changes requested):
@dhruvipatel14 thank you for this, I am yet to double check the meeting minutes for this but for now I would like us to assume we are using `[` without the modifier and hence not a global shortcut. This is easy to change later.

The main issue to resolve is that this does not appear to work when the viewport is small (aka mobile width), due to the way the keyboard controllers override the `bind` key.

Here's a recommended approach, but feel free to suggest something different.

## 1. Update the `KeybaordController` to support a `target`

So that we can easily change the 'button' that we activate, let's add support for a target in the `KeyboardController`.

Below is the full code, notice we now have a `static targets = [...` declared and use the `this.elementTarget` in two places throughout the code. 

You can read more about targets here https://stimulus.hotwired.dev/reference/targets

In addition, there's a small bug we should fix in the `keyValueChanged` where we should not try to bind new things if the key is empty. Plus, let's refine the JSDoc a bit here.

```ts
import { Controller } from '@hotwired/stimulus';
import Mousetrap from 'mousetrap';

// import with side-effect to add global-bind plugin (see https://github.com/ccampbell/mousetrap/tree/master/plugins/global-bind)
import 'mousetrap/plugins/global-bind/mousetrap-global-bind';

/**
 * Adds the ability to trigger a button click event using a
 * keyboard shortcut declared on the controlled element.
 *
 * @see https://craig.is/killing/mice
 *
 * @example
 * ```html
 * <button type="button" data-controller="w-kbd" data-w-kbd-key="[">Trigger me with the <kbd>[</kbd> key.</button>
 * ```
 *
 * @example - use `mod` for `ctrl` on Windows and `cmd` on MacOS
 * ```html
 * <button type="button" data-controller="w-kbd" data-w-kbd-key="mod+s">Trigger me with <kbd>ctrl+p</kbd> on Windows or <kbd>cmd+p</kbd> on MacOS.</button>
 * ```
 *
 * @example - use 'global' scope to allow the shortcut to work even when an input is focused
 * ```html
 * <button type="button" data-controller="w-kbd" data-w-kbd-key="mod+s" data-w-kbd-scope-value="global">Trigger me globally.</button>
 * ```
 *
 * @example - use aria-keyshortcuts (when the key string is compatible with Mousetrap's syntax)
 * @see https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-keyshortcuts
 * ```html
 * <button type="button" data-controller="w-kbd" aria-keyshortcuts="alt+p">Trigger me with alt+p.</button>
 * ```
 */
export class KeyboardController extends Controller<
  HTMLButtonElement | HTMLAnchorElement
> {
  static targets = ['element'];

  static values = {
    key: { default: '', type: String },
    scope: { default: '', type: String },
  };

  /** Keyboard shortcut string. */
  declare keyValue: string;
  /** Scope of the keyboard shortcut, defaults to the normal MouseTrap (non-input) scope. */
  declare scopeValue: '' | 'global';

  /**
   * Allow an explicit target to be the element that the keyboard activates,
   * if not provided, defaults to the controller's element.
   */
  declare readonly elementTarget: HTMLButtonElement | HTMLAnchorElement;
  /** Whether the element target exists. */
  declare readonly hasElementTarget: boolean;

  /**
   * Set up the handleKey binding and check for an aria-keyshortcuts attribute to
   * use as a fallback if the key value is not set.
   */
  initialize() {
    this.handleKey = this.handleKey.bind(this);
    if (!this.keyValue) {
      const ariaKeyShortcuts = (
        this.hasElementTarget ? this.elementTarget : this.element
      ).getAttribute('aria-keyshortcuts');
      if (ariaKeyShortcuts) {
        this.keyValue = ariaKeyShortcuts;
      }
    }
  }

  handleKey(event: Event) {
    if (event.preventDefault) event.preventDefault();
    (this.hasElementTarget ? this.elementTarget : this.element).click();
  }

  /**
   * When a key is set or changed, bind the handler to the keyboard shortcut.
   * This will override the shortcut, if already set.
   * Allow for key to be empty, so it can be removed from the attributes
   * to unbind the shortcut.
   */
  keyValueChanged(key: string, previousKey: string) {
    if (previousKey && previousKey !== key) {
      Mousetrap.unbind(previousKey);
    }

    if (!key) return;

    if (this.scopeValue === 'global') {
      Mousetrap.bindGlobal(key, this.handleKey);
    } else {
      Mousetrap.bind(key, this.handleKey);
    }
  }
}

```

### 2. Update `wagtail/admin/templates/wagtailadmin/base.html` to be the controlled element

We need to then find the parent of the sidebar, and allow this to be the Stimulus controlled element.

```html
    <aside id="wagtail-sidebar" class="sidebar-loading" aria-label="{% trans 'Sidebar' %}" data-wagtail-sidebar data-controller="w-kbd" data-w-kbd-key-value="["></aside>
```

Notice we put the data-controller and the `key-value` here instead of on the buttons.


### 3. Update `client/src/components/Sidebar/Sidebar.tsx`

Instead of the `data-controller` being on each button, we can just put the `data-w-kbd-target=....` on the buttons.

However, we will need to make sure the right button is the target element, with some logic in React we can do this.

I have added individual comments in the code about this.


### 4. Update unit tests

Please ensure we add a new unit test in `client/src/controllers/KeyboardController.test.js` to cover this.

Here's a starting point...

``js

  describe('using an element target instead of the controlled element', () => {
    it('should call the click event when the `j` key is pressed after being registered', async () => {
      expect(buttonClickMock).not.toHaveBeenCalled();

      await setup(
        `<aside data-controller="w-kbd" data-w-kbd-key-value="j"><button id="btn" data-w-kbd-target="element">Go</button></aside>`,
      );

      simulateKey({ key: 'j' }, document.getElementById('btn'));

      expect(buttonClickMock).toHaveBeenCalledTimes(1);
      expect(buttonClickMock.mock.contexts).toEqual([
        document.getElementById('btn'),
      ]);
    });

    /** @TODO - add another for reading the aria shortcuts from the target element */
  });
```

Thank you again, I hope this feedback is OK and all makes sense.

I would love to see this in, sorry it's not as simple as we first thought.

Let us know if you are OK to continue on this PR or if we should close and add these comments to the issue/discussion for someone else to pick up.

dhruvipatel14 commented:
@lb-  Thanks for the detailed feedback.  I appreciate the insights and will work on the suggested changes.

dhruvipatel14 commented:
Hi @lb- ,  I have updated toggle sidebar shortcut to '['.
With updated changes the shortcuts seems to be working fine for smaller viewports with target element. I have also updated tests. Let me know if any additional changes are needed.

lb- reviewed (approved):
@dhruvipatel14  I have just forced push a refinement to this PR that updates the keyboard shortcut to align with the minimap one. I have also rebased your commits & resolved the conflicts, note that you will need to be careful when pulling this branch down in your local machine.

Aside from a small tweak to the JS controller, this is working great locally.

I will not merge though, I will let @thibaudcolas review and see if this should go in 7.1 or hold for further discussions.

cc @Scotchester
